pkg/eventpersisterdwh/persister/persister.go

```diff
@@ -145,25 +174,88 @@ func NewPersisterDWH(
 	}
 	ctx, cancel := context.WithCancel(context.Background())
 	return &PersisterDWH{
-		puller: puller.NewRateLimitedPuller(p, dopts.maxMPS),
-		logger: dopts.logger.Named("persister"),
-		ctx:    ctx,
-		cancel: cancel,
-		doneCh: make(chan struct{}),
-		writer: writer,
-		opts:   dopts,
+		client:                       client,
+		topic:                        topic,
+		subscription:                 subscription,
+		pullerNumGoroutines:          pullerNumGoroutines,
+		pullerMaxOutstandingMessages: pullerMaxOutstandingMessages,
+		pullerMaxOutstandingBytes:    pullerMaxOutstandingBytes,
+		mysqlClient:                  mysqlClient,
+		logger:                       dopts.logger.Named("persister"),
+		ctx:                          ctx,
+		cancel:                       cancel,
+		doneCh:                       make(chan struct{}),
+		writer:                       writer,
+		opts:                         dopts,
 	}
 }
 
 func (p *PersisterDWH) Run() error {
 	defer close(p.doneCh)
-	p.group.Go(func() error {
-		return p.puller.Run(p.ctx)
-	})
-	for i := 0; i < p.opts.numWorkers; i++ {
-		p.group.Go(p.batch)
+	timer := time.NewTimer(p.opts.checkInterval)
+	defer timer.Stop()
+	subscription := make(chan struct{})
+	go p.subscribe(subscription)
+	for {
+		select {
+		case <-timer.C:
+			// check if there are running experiment
+			exist, err := p.checkRunningExperiments(p.ctx)
+			if err != nil {
+				p.logger.Error("Failed to check experiments existence", zap.Error(err))
+				continue
+			}
+			if exist {
+				p.logger.Debug("There are running experiments")
+				if !p.IsRunning() {
+					p.group = errgroup.Group{}
```

Comment:
This should be below `err = p.createNewPuller()`. Otherwise, if it fails to create the puller, it will leave the `p.group` set.